crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1694,27 +1695,43 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
+    /// Causes crashes in the corpus test on:
+    /// - `crates/red_knot_workspace/resources/test/corpus/51_gen_comp2.py`
+    /// - `crates/red_knot_workspace/resources/test/corpus/52_gen_comp_if.py`
+    /// - `crates/red_knot_workspace/resources/test/corpus/53_dict_comp.py`
     fn infer_comprehension_definition(
         &mut self,
         iterable: &ast::Expr,
         target: &ast::ExprName,
         is_first: bool,
+        is_async: bool,
         definition: Definition<'db>,
     ) {
-        if !is_first {
-            let expression = self.index.expression(iterable);
-            let result = infer_expression_types(self.db, expression);
-            self.extend(result);
-            let _iterable_ty = self
-                .types
-                .expression_ty(iterable.scoped_ast_id(self.db, self.scope));
-        }
-        // TODO(dhruvmanila): The iter type for the first comprehension is coming from the
-        // enclosing scope.
+        let expression = self.index.expression(iterable);
+        let result = infer_expression_types(self.db, expression);
+        self.extend(result);
+
+        let lookup_scope = if is_first {
+            self.index
+                .parent_scope_id(self.scope.file_scope_id(self.db))
+                .expect("A comprehension should never be the top-level scope")
+                .to_scope_id(self.db, self.file)
```

Comment:
It might not be the first parent in case there are nested comprehensions like `[[x for x in iter1] for y in iter2]`

Reply:
I just pushed a test for that. It seems to work fine :)

I think the nonlocal name lookup happens automatically now, following https://github.com/astral-sh/ruff/commit/3c4ec82aee18c8438a5f807853edcf703e0816a0. We need to look it up in the parent scope because it is _not_ valid for the iterable to reference any variables defined in _this_ scope. But looking it up in the parent scope doesn't mean that it will stop at the parent scope; it will continue iterating up through the parent scopes until it finds a definition (or infer `Unbound` if it can't)

Reply:
Oh interesting, thanks for clarifying